src/google/cloud/ndb/_gql.py

```diff
@@ -0,0 +1,761 @@
+            parameters (list): list of bound parameters passed to 'operator'
+                before creating the filter. When using the parameters as a
+                pass-through, pass 'nop' into the operator field and the first
+                value will be used unprocessed).
+
+        Returns:
+            True: if the filter was okay to add.
+        """
+        if parameters[0] is None:
+            return False
+
+        filter_rule = (identifier, condition)
+        if identifier.lower() == "ancestor":
+            self._has_ancestor = True
+            filter_rule = (self._ANCESTOR, "is")
+            assert condition.lower() == "is"
+
+        if operator == "list" and condition.lower() != "in":
+            self._Error("Only IN can process a list of values")
+
+        self._filters.setdefault(filter_rule, []).append(
+            (operator, parameters)
+        )
+        return True
+
+    def _AddSimpleFilter(self, identifier, condition, parameter):
+        """Add a filter to the query being built (no post-processing on parameter).
+
+        Args:
+            identifier (str): identifier being used in comparison.
+            condition (str): comparison operator used in the filter.
+            parameter (Union[str, int, Literal]: ID of the reference being made
+                or a value of type Literal
+
+        Returns:
+            bool: True if the filter could be added. False otherwise.
+        """
+        return self._AddProcessedParameterFilter(
+            identifier, condition, "nop", [parameter]
+        )
+
+    def _Identifier(self):
+        """Consume an identifier and return it.
+
+        Returns:
+            str: The identifier string. If quoted, the surrounding quotes are
+                stripped.
+        """
+        identifier = self._AcceptRegex(self._identifier_regex)
+        if identifier:
+            if identifier.upper() in self.RESERVED_KEYWORDS:
+                self._next_symbol -= 1
+                self._Error("Identifier is a reserved keyword")
+        else:
+            identifier = self._AcceptRegex(self._quoted_identifier_regex)
+            if identifier:
+                identifier = identifier[1:-1].replace('""', '"')
+        return identifier
+
+    def _ExpectIdentifier(self):
+        id = self._Identifier()
+        if not id:
+            self._Error("Identifier Expected")
+        return id
+
+    def _Reference(self):
+        """Consume a parameter reference and return it.
+
+        Consumes a reference to a positional parameter (:1) or a named parameter
+            (:email). Only consumes a single reference (not lists).
+
+        Returns:
+            Union[str, int]: The name of the reference (integer for positional
+                parameters or string for named parameters) to a bind-time
+                parameter.
+        """
+        reference = self._AcceptRegex(self._ordinal_regex)
+        if reference:
+            return int(reference)
```

Comment:
Why does this cast reference to int but the version a few lines down using the named regex doesn't? Some nuance on expected result form _AcceptRegex?

Reply:
Yeah, _AcceptRegex returns a string, which is fine for the named regex (e.g. WHERE col=:some_name). However, ordinal regex is meant to return an int (e.g. WHERE col=:1), so that's why there's a cast.